src/input/return_enums.rs

```diff
@@ -371,7 +378,10 @@ impl<'a> GenericIterable<'a> {
         field_type: &'static str,
         max_length: Option<usize>,
     ) -> ValResult<'a, Vec<PyObject>> {
-        let max_length_check = MaxLengthCheck::new(max_length, field_type, input);
+        let capacity = self
+            .generic_len()
+            .unwrap_or_else(|| max_length.unwrap_or(DEFAULT_CAPACITY));
```

Comment:
That's right, for the first part but for the iterator I realise it wasn't working. The max_length need to be caught by a None condition. So if someone validates this:

```
from pydantic_core import SchemaValidator


def infinite_generator():
    i = 0
    while True:
        yield i
        i += 1


v = SchemaValidator({'type': 'list'})
v.validate_python(infinite_generator())
```
Then the error message will be:
<img width="749" alt="Screenshot 2023-08-23 at 21 10 27" src="https://github.com/pydantic/pydantic-core/assets/31091198/85afd7dd-f3d8-480f-8cb2-44dca574b368">

Otherwise this will run into an infinite loop.